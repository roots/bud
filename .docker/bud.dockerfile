FROM node:lts

LABEL name 'bud'
LABEL version 1

RUN npm install -g verdaccio pm2

COPY ./.docker/bud/motd /etc/motd
COPY ./.docker/bud/bash.bashrc /etc/bash.bashrc
