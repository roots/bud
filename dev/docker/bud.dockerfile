FROM node:16

LABEL name 'bud'
LABEL version 1

RUN npm install netlify-cli npm-cli-login --global

COPY ./dev/docker/bud/motd /etc/motd
COPY ./dev/docker/bud/bash.bashrc /etc/bash.bashrc

COPY ./ /srv/bud

RUN mkdir -p /srv/bud/node_modules \
  && mkdir -p /srv/bud/storage/verdaccio \
  && mkdir -p /srv/bud/storage/yarn \
  && mkdir -p /srv/bud/storage/coverage \
  && mkdir -p /srv/mocks
