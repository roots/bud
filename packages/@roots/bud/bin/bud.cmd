@echo off

node "%~dp0\bud.js" %*
