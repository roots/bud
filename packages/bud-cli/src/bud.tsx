#!/usr/bin/env node
import {yargs} from '@roots/bud-support'
import * as build from './commands/build'
import * as publish from './commands/publish'

/**
 * $bud
 */
const cli = yargs
  .command(build)
  .command(publish)
  .recommendCommands()
  .demandCommand(1, 'Try using one of the above commands.')
  .usage('bud [command] [options]')
  .version()
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(true)
  .epilog('https://github.com/roots/bud').argv

export {cli as bud}
module.exports = {bud: cli}
