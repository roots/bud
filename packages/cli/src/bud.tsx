#!/usr/bin/env node

import {yargs} from '@roots/bud-support'
import * as build from './commands/build'
import * as publish from './commands/publish'

module.exports = yargs
  .command(build)
  .command(publish)
  .recommendCommands()
  .demandCommand(1, 'Try using one of the above commands.')
  .usage('bud [command] [options]')
  .version()
  .wrap(yargs.terminalWidth())
  .showHelpOnFail(true)
  .epilog('https://github.com/roots/bud').argv
