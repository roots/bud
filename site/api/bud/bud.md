---
id: bud.bud
title: bud namespace
sidebar_label: bud namespace
hide_title: true
sidebar: "api"
slug: bud
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## Bud namespace

Signature:

```typescript
export declare namespace Bud
```

## Type Aliases

| Type Alias                      | Description             |
| ------------------------------- | ----------------------- |
| [Options](/api/bud/bud/options) | Bud constructor options |
