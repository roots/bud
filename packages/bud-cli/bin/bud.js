#!/usr/bin/node
require('./../lib/bud')
