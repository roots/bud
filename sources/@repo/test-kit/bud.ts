import * as CONSTANTS from '@repo/constants'
import {Bud, factory as budFactory, makeContext, seed} from '@roots/bud'
import {Config} from '@roots/bud-framework'
import {join} from 'node:path'

export const repoPath = (...path: Array<string>) =>
  join(CONSTANTS.REPO_PATH, ...(path ?? []))

export const mockProject = {
  path: repoPath('tests/util/project'),
}

export const factory = async (options?: Config.Options) => {
  const context = await makeContext(repoPath('tests/util/project'))

  const bud = await budFactory({
    name: 'bud',
    mode: 'production',
    ...(options ?? {}),
    context: {
      ...context,
      ...(options?.context ?? {}),
      args: {
        ...(options?.context?.args ?? {}),
        ci: true,
      },
    },
    seed: {
      ...seed,
      ...(options?.seed ?? {}),
    },
  })

  return bud
}

export {Bud}
