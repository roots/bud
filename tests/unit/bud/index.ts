import {Bud, config, factory, Framework} from '@roots/bud'

describe('factory', () => {
  let bud: Framework

  let snapshotConfig = {
    build: {
      bail: true,
      devtool: false,
      experiments: {
        lazyCompilation: false,
      },
      infrastructureLogging: {
        appendOnly: true,
        level: 'none',
      },
      optimization: {
        emitOnErrors: false,
        minimizer: ['...'],
        moduleIds: 'deterministic',
        removeEmptyChunks: true,
        splitChunks: {
          cacheGroups: {
            vendor: {
              chunks: 'all',
              filename: 'vendor/[name].js',
              priority: -10,
              reuseExistingChunk: true,
              test: /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
            },
          },
        },
      },
      output: {
        pathinfo: false,
      },
      performance: {},
      profile: false,
      resolve: {
        extensions: [
          '.wasm',
          '.mjs',
          '.js',
          '.jsx',
          '.css',
          '.json',
          '.json5',
          '.toml',
          '.xml',
          '.csv',
          '.tsv',
          '.yml',
          '.yaml',
          '.xml',
        ],
      },
      stats: {},
    },
    ci: true,
    clean: true,
    debug: true,
    discover: false,
    extension: {
      cleanWebpackPlugin: {
        cleanOnceBeforeBuildPatterns: ['**/*', '!dll'],
        cleanStaleWebpackAssets: true,
        protectWebpackAssets: true,
      },
      cssMinimizerWebpackPlugin: {
        minimizerOptions: {
          preset: [
            'default',
            {
              discardComments: {
                removeAll: true,
              },
            },
          ],
        },
      },
      'html-webpack-plugin': {
        alwaysWriteToDisk: true,
        inject: true,
        minify: {
          collapseWhitespace: false,
          keepClosingSlash: true,
          removeComments: true,
          removeRedundantAttributes: true,
          removeScriptTypeAttributes: true,
          removeStyleLinkTypeAttributes: true,
          useShortDoctype: true,
        },
      },
      'interpolate-html-plugin': {
        replace: {},
      },
      miniCssExtractPlugin: {},
      webpackConfigDumpPlugin: {
        depth: 8,
        keepCircularReferences: true,
      },
      webpackDefinePlugin: {},
      webpackManifestPlugin: {
        fileName: 'manifest.json',
        writeToFileEmit: true,
      },
      webpackProvidePlugin: {},
    },
    fileFormat: '[name]',
    hash: false,
    hashFormat: '[name].[contenthash:6]',
    html: false,
    install: false,
    location: {
      dist: 'dist',
      modules: 'node_modules',
      project: expect.any(String),
      publicPath: '',
      src: 'src',
      storage: '.budfiles',
    },
    log: false,
    manifest: true,
    minimize: true,
    name: 'bud',
    patterns: {
      css: /\\\\\\.css\\$/,
      cssModule: /\\\\\\.module\\\\\\.css\\$/,
      csv: /\\\\\\.\\(csv\\|tsv\\)\\$/,
      font: /\\\\\\.\\(ttf\\|otf\\|eot\\|woff2\\?\\|ico\\)\\$/,
      html: /\\\\\\.\\(html\\?\\)\\$/,
      image: /\\\\\\.\\(png\\|jpe\\?g\\|gif\\)\\$/,
      js: /\\\\\\.\\(js\\|jsx\\)\\$/,
      json: /\\\\\\.json\\$/,
      json5: /\\\\\\.json5\\$/,
      md: /\\\\\\.md\\$/,
      modules: /\\(node_modules\\|bower_components\\)/,
      sass: /\\\\\\.\\(scss\\|sass\\)\\$/,
      sassModule: /\\\\\\.module\\\\\\.\\(scss\\|sass\\)\\$/,
      svg: /\\\\\\.svg\\$/,
      toml: /\\\\\\.toml\\$/,
      ts: /\\\\\\.\\(ts\\|tsx\\)\\$/,
      vue: /\\\\\\.vue\\$/,
      xml: /\\\\\\.xml\\$/,
      yml: /\\\\\\.\\(yaml\\|yml\\)\\$/,
    },
    server: {
      browser: {
        indicator: true,
        log: true,
        overlay: true,
      },
      host: 'localhost',
      methods: ['GET', 'HEAD'],
      middleware: {
        dev: true,
        hot: true,
        proxy: false,
      },
      port: 3000,
      proxy: {
        host: 'localhost',
        port: 8000,
      },
      watch: {
        files: [],
        options: {},
      },
    },
    theme: {
      colors: {
        accent: '#ff69b4',
        error: '#dc3545',
        errorAlt: '#b22222',
        faded: '#6C758F',
        flavor: '#78C5D7',
        foreground: '#FFFFFF',
        primary: '#545DD7',
        primaryAlt: '#663399',
        success: '#46D46A',
        warning: '#FF611A',
      },
      columns: 12,
      maxHeight: 999,
      maxWidth: 80,
      screens: [
        [0, 40],
        [41, 60],
        [61, 80],
        [81, 200],
      ],
      spacing: 1,
    },
  }

  beforeAll(() => {
    bud = factory({
      config: {...config, ci: true},
    })
  })

  afterAll(done => {
    bud.close(done)
  })

  it('is a function', () => {
    expect(factory).toBeInstanceOf(Function)
  })

  it('makes a new instace of bud', () => {
    expect(bud).toBeInstanceOf(Bud)
  })

  it('config matches snapshot', () => {
    expect(config).toMatchSnapshot(snapshotConfig)
  })

  it('bud.store matches snapshot', () => {
    expect(bud.store.all()).toMatchSnapshot({
      ...snapshotConfig,
      ci: true,
    })
  })
})
