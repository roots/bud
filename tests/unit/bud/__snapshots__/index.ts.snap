// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory bud.store matches snapshot 1`] = `
Object {
  "build": Object {
    "bail": true,
    "devtool": false,
    "experiments": Object {
      "lazyCompilation": false,
    },
    "infrastructureLogging": Object {
      "appendOnly": true,
      "level": "none",
    },
    "node": false,
    "optimization": Object {
      "emitOnErrors": false,
      "minimizer": Array [
        "...",
      ],
      "moduleIds": "deterministic",
      "removeEmptyChunks": true,
      "splitChunks": Object {
        "cacheGroups": Object {
          "vendor": Object {
            "chunks": "all",
            "filename": "vendor/[name].js",
            "priority": -10,
            "reuseExistingChunk": true,
            "test": Object {},
          },
        },
      },
    },
    "output": Object {
      "pathinfo": false,
    },
    "performance": Object {},
    "profile": false,
    "resolve": Object {
      "extensions": Array [
        ".wasm",
        ".mjs",
        ".js",
        ".jsx",
        ".css",
        ".json",
        ".json5",
        ".toml",
        ".xml",
        ".csv",
        ".tsv",
        ".yml",
        ".yaml",
        ".xml",
      ],
    },
    "stats": Object {},
  },
  "ci": true,
  "clean": true,
  "debug": true,
  "discover": false,
  "extension": Object {
    "cleanWebpackPlugin": Object {
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
        "!dll",
      ],
      "cleanStaleWebpackAssets": true,
      "protectWebpackAssets": true,
    },
    "cssMinimizerWebpackPlugin": Object {
      "minimizerOptions": Object {
        "preset": Array [
          "default",
          Object {
            "discardComments": Object {
              "removeAll": true,
            },
          },
        ],
      },
    },
    "html-webpack-plugin": Object {
      "alwaysWriteToDisk": true,
      "inject": true,
      "minify": Object {
        "collapseWhitespace": false,
        "keepClosingSlash": true,
        "removeComments": true,
        "removeRedundantAttributes": true,
        "removeScriptTypeAttributes": true,
        "removeStyleLinkTypeAttributes": true,
        "useShortDoctype": true,
      },
    },
    "interpolate-html-plugin": Object {
      "replace": Object {},
    },
    "miniCssExtractPlugin": Object {},
    "webpackConfigDumpPlugin": Object {
      "depth": 8,
      "keepCircularReferences": true,
    },
    "webpackDefinePlugin": Object {},
    "webpackManifestPlugin": Object {
      "fileName": "manifest.json",
      "writeToFileEmit": true,
    },
    "webpackProvidePlugin": Object {},
  },
  "fileFormat": "[name]",
  "hash": false,
  "hashFormat": "[name].[contenthash:6]",
  "html": false,
  "install": false,
  "location": Object {
    "dist": "dist",
    "modules": "node_modules",
    "project": Any<String>,
    "publicPath": "",
    "src": "src",
    "storage": ".budfiles",
  },
  "log": false,
  "manifest": true,
  "minimize": true,
  "name": "bud",
  "patterns": Object {
    "css": Object {},
    "cssModule": Object {},
    "csv": Object {},
    "font": Object {},
    "html": Object {},
    "image": Object {},
    "js": Object {},
    "json": Object {},
    "json5": Object {},
    "md": Object {},
    "modules": Object {},
    "sass": Object {},
    "sassModule": Object {},
    "svg": Object {},
    "toml": Object {},
    "ts": Object {},
    "vue": Object {},
    "xml": Object {},
    "yml": Object {},
  },
  "server": Object {
    "browser": Object {
      "indicator": true,
      "log": true,
      "overlay": true,
    },
    "host": "localhost",
    "methods": Array [
      "GET",
      "HEAD",
    ],
    "middleware": Object {
      "dev": true,
      "hot": true,
      "proxy": false,
    },
    "port": 3000,
    "proxy": Object {
      "host": "localhost",
      "port": 8000,
    },
    "watch": Object {
      "files": Array [],
      "options": Object {},
    },
  },
  "theme": Object {
    "colors": Object {
      "accent": "#ff69b4",
      "error": "#dc3545",
      "errorAlt": "#b22222",
      "faded": "#6C758F",
      "flavor": "#78C5D7",
      "foreground": "#FFFFFF",
      "primary": "#545DD7",
      "primaryAlt": "#663399",
      "success": "#46D46A",
      "warning": "#FF611A",
    },
    "columns": 12,
    "maxHeight": 999,
    "maxWidth": 80,
    "screens": Array [
      Array [
        0,
        40,
      ],
      Array [
        41,
        60,
      ],
      Array [
        61,
        80,
      ],
      Array [
        81,
        200,
      ],
    ],
    "spacing": 1,
  },
}
`;

exports[`factory config matches snapshot 1`] = `
Object {
  "build": Object {
    "bail": true,
    "devtool": false,
    "experiments": Object {
      "lazyCompilation": false,
    },
    "infrastructureLogging": Object {
      "appendOnly": true,
      "level": "none",
    },
    "node": false,
    "optimization": Object {
      "emitOnErrors": false,
      "minimizer": Array [
        "...",
      ],
      "moduleIds": "deterministic",
      "removeEmptyChunks": true,
      "splitChunks": Object {
        "cacheGroups": Object {
          "vendor": Object {
            "chunks": "all",
            "filename": "vendor/[name].js",
            "priority": -10,
            "reuseExistingChunk": true,
            "test": Object {},
          },
        },
      },
    },
    "output": Object {
      "pathinfo": false,
    },
    "performance": Object {},
    "profile": false,
    "resolve": Object {
      "extensions": Array [
        ".wasm",
        ".mjs",
        ".js",
        ".jsx",
        ".css",
        ".json",
        ".json5",
        ".toml",
        ".xml",
        ".csv",
        ".tsv",
        ".yml",
        ".yaml",
        ".xml",
      ],
    },
    "stats": Object {},
  },
  "ci": true,
  "clean": true,
  "debug": true,
  "discover": false,
  "extension": Object {
    "cleanWebpackPlugin": Object {
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
        "!dll",
      ],
      "cleanStaleWebpackAssets": true,
      "protectWebpackAssets": true,
    },
    "cssMinimizerWebpackPlugin": Object {
      "minimizerOptions": Object {
        "preset": Array [
          "default",
          Object {
            "discardComments": Object {
              "removeAll": true,
            },
          },
        ],
      },
    },
    "html-webpack-plugin": Object {
      "alwaysWriteToDisk": true,
      "inject": true,
      "minify": Object {
        "collapseWhitespace": false,
        "keepClosingSlash": true,
        "removeComments": true,
        "removeRedundantAttributes": true,
        "removeScriptTypeAttributes": true,
        "removeStyleLinkTypeAttributes": true,
        "useShortDoctype": true,
      },
    },
    "interpolate-html-plugin": Object {
      "replace": Object {},
    },
    "miniCssExtractPlugin": Object {},
    "webpackConfigDumpPlugin": Object {
      "depth": 8,
      "keepCircularReferences": true,
    },
    "webpackDefinePlugin": Object {},
    "webpackManifestPlugin": Object {
      "fileName": "manifest.json",
      "writeToFileEmit": true,
    },
    "webpackProvidePlugin": Object {},
  },
  "fileFormat": "[name]",
  "hash": false,
  "hashFormat": "[name].[contenthash:6]",
  "html": false,
  "install": false,
  "location": Object {
    "dist": "dist",
    "modules": "node_modules",
    "project": Any<String>,
    "publicPath": "",
    "src": "src",
    "storage": ".budfiles",
  },
  "log": false,
  "manifest": true,
  "minimize": true,
  "name": "bud",
  "patterns": Object {
    "css": Object {},
    "cssModule": Object {},
    "csv": Object {},
    "font": Object {},
    "html": Object {},
    "image": Object {},
    "js": Object {},
    "json": Object {},
    "json5": Object {},
    "md": Object {},
    "modules": Object {},
    "sass": Object {},
    "sassModule": Object {},
    "svg": Object {},
    "toml": Object {},
    "ts": Object {},
    "vue": Object {},
    "xml": Object {},
    "yml": Object {},
  },
  "server": Object {
    "browser": Object {
      "indicator": true,
      "log": true,
      "overlay": true,
    },
    "host": "localhost",
    "methods": Array [
      "GET",
      "HEAD",
    ],
    "middleware": Object {
      "dev": true,
      "hot": true,
      "proxy": false,
    },
    "port": 3000,
    "proxy": Object {
      "host": "localhost",
      "port": 8000,
    },
    "watch": Object {
      "files": Array [],
      "options": Object {},
    },
  },
  "theme": Object {
    "colors": Object {
      "accent": "#ff69b4",
      "error": "#dc3545",
      "errorAlt": "#b22222",
      "faded": "#6C758F",
      "flavor": "#78C5D7",
      "foreground": "#FFFFFF",
      "primary": "#545DD7",
      "primaryAlt": "#663399",
      "success": "#46D46A",
      "warning": "#FF611A",
    },
    "columns": 12,
    "maxHeight": 999,
    "maxWidth": 80,
    "screens": Array [
      Array [
        0,
        40,
      ],
      Array [
        41,
        60,
      ],
      Array [
        61,
        80,
      ],
      Array [
        81,
        200,
      ],
    ],
    "spacing": 1,
  },
}
`;
